File: R/p3.r
# The prime factors of 13195 are 5, 7, 13 and 29.

# What is the largest prime factor of the number 600851475143 ?

#  The largest factor must be less than:
max_factor <- as.integer(ceiling(sqrt(600851475143)))
#  Since 600851475143 isn't even, only check odd numbers:
potential_factors <- seq(max_factor, 1, by = -2)


#  Which of potential_factors are prime?
is_prime <- function(x){
	max_check <- as.integer(ceiling(sqrt(x)))
	if(x > 4){
		to_check <- c(2, seq(3, max_check, by = 2))
	} else {
		to_check <- c(2)
	}
	
	rval <- all(x %% to_check != 0)
	return(rval)
}
system.time(prime_factors <- sapply(potential_factors, is_prime))
factors <- potential_factors[prime_factors]


#  initialize search of prime factors
n_search <- length(factors)
largest_factor <- NA
found_largest <- FALSE
n <- 1
while(found_largest == FALSE){
	if(as.numeric(600851475143.0) %% factors[n] == 0){
		largest_factor <- factors[n]
		found_largest <- TRUE
	} else {
		n <- n + 1
	}
}
largest_factor

File: Python/p3.py
# The prime factors of 13195 are 5, 7, 13 and 29.

# What is the largest prime factor of the number 600851475143 ?

#  math libs:
from math import ceil, sqrt

#  Time code:
import time
start_time = time.clock()

#  Function to check whether a number is prime (simple exhaustive search)
def is_prime(x):
	#  largest number to check for divisibility 
	max_int = int(ceil(sqrt(x))) + 1
	
	#  Search for a factor, stop early if one is found
	search_range = [2] + range(3, max_int, 2)
	rval = True
	if(x != 2 and x != 3):
		for i in search_range:
			if x % i == 0:
				rval = False
				break
	#  return bool
	return rval


#  Find largest prime factor. Method:
#   1. Seach for a factor (descending)
#   2. Once a factor is found, check if prime. If not, continue search for next factor
def find_factor(y):
	max_int = int(ceil(sqrt(y)))
	
	#  if y is odd, only search odd numbers
	if(y % 2 != 0):
		search_range = [2] + range(3, max_int, 2)
	else:
		search_range = range(2, max_int)
	
	#  Search
	rval = 0
	for k in reversed(search_range):
		if y % k == 0:
			if is_prime(k):
				rval = k
				break
	return rval

print find_factor(600851475143)

end_time = time.clock()
print("Elapsed time was:  %g seconds" % (end_time - start_time))
	
# for j in range(1, 10):
	# print '\nis %d prime?\t' % j, is_prime(j)
		